### PR TITLE
Use backing fields when available

### PR DESCRIPTION
**Pull request by MarcinJuraszek**
It's a simplified version of what @deipax did in #22.

The way we're doing per-type clone delegate caching makes additional caching of reflection data unnecessary. So instead of using `ModelInfo<T>` from `Deipax.Core.Common` it's simply checking the fields when the clone delegate is generated, without any additional caching.

Benchmark results:

**.NET Core - before**
```
                                      Method |           Mean |         Error |        StdDev |
-------------------------------------------- |---------------:|--------------:|--------------:|
                         GetCloneSimpleClass |       132.5 ns |      1.793 ns |      1.296 ns |
                          GetCloneListOfInts |     4,751.3 ns |     67.455 ns |     52.664 ns |
       GetCloneListOfSimpleClassSameInstance |   632,354.0 ns |  8,291.360 ns |  7,350.074 ns |
 GetCloneListOfSimpleClassDifferentInstances | 1,806,829.9 ns | 27,005.394 ns | 23,939.576 ns |
```
**.NET Core - with #22 applied**
```
                                      Method |           Mean |         Error |        StdDev |
-------------------------------------------- |---------------:|--------------:|--------------:|
                         GetCloneSimpleClass |       133.7 ns |      2.630 ns |      3.131 ns |
                          GetCloneListOfInts |     4,721.2 ns |     91.612 ns |    105.500 ns |
       GetCloneListOfSimpleClassSameInstance |   651,762.3 ns | 10,653.091 ns |  8,895.813 ns |
 GetCloneListOfSimpleClassDifferentInstances | 1,770,150.1 ns | 31,561.661 ns | 27,978.587 ns |
```

**.NET Core - with these changes**
```
                                      Method |           Mean |         Error |        StdDev |
-------------------------------------------- |---------------:|--------------:|--------------:|
                         GetCloneSimpleClass |       129.1 ns |      1.655 ns |      1.548 ns |
                          GetCloneListOfInts |     4,706.9 ns |     87.409 ns |     72.991 ns |
       GetCloneListOfSimpleClassSameInstance |   665,777.9 ns |  9,338.243 ns |  8,278.109 ns |
 GetCloneListOfSimpleClassDifferentInstances | 1,811,359.6 ns | 29,407.678 ns | 26,069.138 ns |
```
**.NET Framework 4.6.1 - before**
```
                                      Method |           Mean |          Error |         StdDev |
-------------------------------------------- |---------------:|---------------:|---------------:|
                         GetCloneSimpleClass |       255.2 ns |       3.486 ns |       2.911 ns |
                          GetCloneListOfInts |     4,683.8 ns |      60.154 ns |      50.231 ns |
       GetCloneListOfSimpleClassSameInstance | 1,117,613.7 ns |  20,728.126 ns |  17,308.923 ns |
 GetCloneListOfSimpleClassDifferentInstances | 5,182,728.0 ns | 110,875.858 ns | 123,238.212 ns |
```
**.NET Framework 4.6.1 - with #22 applied**
```
                                      Method |           Mean |         Error |        StdDev |
-------------------------------------------- |---------------:|--------------:|--------------:|
                         GetCloneSimpleClass |       149.7 ns |      3.023 ns |      7.359 ns |
                          GetCloneListOfInts |     4,691.6 ns |     71.216 ns |     55.600 ns |
       GetCloneListOfSimpleClassSameInstance |   939,968.9 ns | 17,599.375 ns | 16,462.464 ns |
 GetCloneListOfSimpleClassDifferentInstances | 2,385,342.2 ns | 71,420.279 ns | 59,639.162 ns |
```
**.NET Framework 4.6.1 - with these changes**
```
                                      Method |           Mean |         Error |        StdDev |
-------------------------------------------- |---------------:|--------------:|--------------:|
                         GetCloneSimpleClass |       138.1 ns |      1.481 ns |      1.386 ns |
                          GetCloneListOfInts |     4,668.7 ns |     83.819 ns |     74.304 ns |
       GetCloneListOfSimpleClassSameInstance |   971,472.9 ns | 18,936.852 ns | 30,579.487 ns |
 GetCloneListOfSimpleClassDifferentInstances | 2,372,608.3 ns | 42,565.389 ns | 39,815.687 ns |
```

To summarize above
- pretty much no changes for .NET Core
- noticeable performance improvements for .NET Framework
- similar performance of this code and #22

### CODE REVIEW

**Comment by deipax** [comment]
Marcin,
I really like your test output, I would be very interested in learning how you did it. 

I would really like to dig into why there was no performance gain with the .Net Core ... if I had to guess, I would say that it looks like the backing fields either do not exist or were not found.

It has been quite a while since I authored the ModelInfo class, but if I recall correctly, when fetching private fields for a type it only does so for that type and none of the classes from which it inherits.  If a class has deep inheritance, I believe only the backing fields for the top class would be returned the way it is authored now.

Look for - private static IReadOnlyList<IFieldModelInfo> GetAllFields(Type t, int depth = 0) - in the modelInfo class.   It is a recursive function that traverses the inheretance chain and builds the full list of private fields.

Once you have the full list of backing fields, the code you have should be good.

**Comment by MarcinJuraszek** [comment]
Good call. I updated the change with some extra logic to use backing fields for properties from base classes too. I validated that it works. I also validated that backing fields are used for both .NET Core and .NET Framework. I suspect there are some optimizations in .NET Core which make property access same as field access for automatic properties. That's why there is no difference.

The new benchmark results:

**.NET Core**
```
                                      Method |           Mean |         Error |        StdDev |
-------------------------------------------- |---------------:|--------------:|--------------:|
                         GetCloneSimpleClass |       128.2 ns |      2.569 ns |      3.765 ns |
                          GetCloneListOfInts |     4,531.7 ns |     60.566 ns |     56.653 ns |
       GetCloneListOfSimpleClassSameInstance |   636,932.2 ns |  6,268.455 ns |  5,556.821 ns |
 GetCloneListOfSimpleClassDifferentInstances | 1,710,996.7 ns | 17,378.553 ns | 16,255.908 ns |
```
**.NET Framework 4.6.1**
```
                                      Method |           Mean |         Error |        StdDev |
-------------------------------------------- |---------------:|--------------:|--------------:|
                         GetCloneSimpleClass |       135.0 ns |      1.947 ns |      1.821 ns |
                          GetCloneListOfInts |     4,531.9 ns |     57.493 ns |     50.966 ns |
       GetCloneListOfSimpleClassSameInstance |   912,664.9 ns |  3,777.194 ns |  3,154.128 ns |
 GetCloneListOfSimpleClassDifferentInstances | 2,267,383.3 ns | 21,317.802 ns | 18,897.674 ns |
```

These benchmarks are done using BenchmarkDotNet (see http://benchmarkdotnet.org/). It's awesome and I highly recommend it. It's really easy to use, you can see a change where I added them to this repo here: https://github.com/MarcinJuraszek/CloneExtensions/commit/63a438f503a2cc754ad3458dfb34714734626eb7

**Comment by deipax** [comment]
Marcin,
Good work!  That should do it.  The only thing I can see now that may cause an issue would be the this:

`var backingFields = GetBackingFields(_type).ToDictionary(f => f.Name);`

If, for some reason, someone used "new" on a property and named the property the same name to hide the property of a parent, this would throw an exception.  Nit picky, I know, but I thought I should mention it.

I will play with the Benchmark.Net, it looks like a solid tool.  I will revert my changes and sync from your branch to grab your most recent work.

-Jeff

**Comment by MarcinJuraszek** [comment]
That's not nit picking. That's a valid concern. I make a check that the property and backing field are both defined in the same type:

```
if (backingFields.TryGetValue("<" + property.Name + ">k__BackingField", out fieldInfo) && fieldInfo.DeclaringType == property.DeclaringType) 
```

But there is still a potential exception in `ToDictionary`. I'll fix that tonight and merge the changes to master.

**Comment by MarcinJuraszek** [comment]
So, I fixed the exception, but turns out cloning properties that are shadowed doesn't really work that well after all with or without my change. I logged a separate issue for that (https://github.com/MarcinJuraszek/CloneExtensions/issues/25) not to block this change from being merged.
